### PR TITLE
bpo-27645: Supporting native backup facility of SQLite

### PR DESCRIPTION
**Pull request by lelit**
This is a new version of PR #377, rebased on more recent `master` to solve the conflicts on the old `NEWS` file. 

It implements a [new requested feature](https://bugs.python.org/issue27645#msg304789) since last review.

See https://bugs.python.org/issue27645

https://bugs.python.org/issue27645

### CODE REVIEW

**Comment by lelit** [comment]
Compilation [fails](https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.7951#L498) on Windows, due to misdeclaration of `unlink()`. I call that function to remove an incomplete backup when an error happens: I looked around in the sources and couldn't figure out a portable way to do that. Any hint?

File: Modules/_sqlite/connection.c
Comment (tiran) [inline Modules/_sqlite/connection.c]: This won't work under Windows. You have to use ``DeleteFileW``, see posixmodule.c

**Review by tiran — CHANGES_REQUESTED** [review]

**Comment by lelit** [comment]
@tiran Thank you, I tried your hint but something's still wrong: maybe I'm not importing the right header (accordingly to what I found on [msdn](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363915(v=vs.85).aspx) it should be `windows.h`), but the compilation [fails](https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.9093#L507) with an _unresolved symbol_. Any suggestion on what I'm doing wrong?

**Comment by tiran** [comment]
Ah, welcome to Windows :)

First of all you want to set ``WIN32_LEAN_AND_MEAN`` before you include the header file.
```
#ifdef MS_WINDOWS
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#endif
```

Next up you are getting an undefined name, not an undefined symbol: ``'DeleteFileW' undefined; assuming extern returning int``. You got me confused for a second. Undefined names are compiler errors, undefined symbols are linker errors. I'm 99% sure that the solution for your problem is explained in https://docs.python.org/3/c-api/intro.html#include-files

> Since Python may define some pre-processor definitions which affect the standard headers on some systems, you must include Python.h before any standard headers are included.

Finally you cannot use ``DeleteFileW`` like that. It does not take a ``const char*`` but a wide char argument. You also need to handle errors differently. You have to use the correct ``PyErr_SetExcFromWindows*()`` function with a wide char filename to set an exception from a Windows error.

**Comment by tiran** [comment]
You got a segfault on Windows:

```
CheckDatabaseSourceName (sqlite3.test.backup.BackupTests) ... Windows fatal exception: access violation
```

**Comment by lelit** [comment]
Yes, I'm trying to learn how the exception chaining works... not having a Windows around does not help, though.

**Comment by lelit** [comment]
Will try harder to understand the issue: thank you again for the hints, and in advance for further advices :)

**Comment by lelit** [comment]
It seems that I was finally able to fix the Windows issue!

I have made the requested changes; please review again.

File: Lib/sqlite3/test/backup.py
Comment (tiran) [inline Lib/sqlite3/test/backup.py]: Please don't follow the example of the other sqlite tests and use standard test names ``test_foo``.

File: Modules/_sqlite/connection.c
Comment (tiran) [inline Modules/_sqlite/connection.c]: PEP 7, the ``*`` come before the var name, not after the type. It's ``PyObject *progress``. Same for ``char *name``.

File: Modules/_sqlite/connection.c
Comment (tiran) [inline Modules/_sqlite/connection.c]: You have defined a finally target but don't use it to clean up resources. Traditionally Python's C either uses goto to perform cleanups or returns immediately.

File: Doc/library/sqlite3.rst
Comment (tiran) [inline Doc/library/sqlite3.rst]: Almost all CPython functions and methods use seconds as double for timeout. Milliseconds is going to confuse people.

File: Modules/_sqlite/connection.c
Comment (tiran) [inline Modules/_sqlite/connection.c]: See other comment, Format ``d`` accepts ints and floats and converts them to double. Multiple the double with 1000 and cast it to int ``sqlite3_sleep(sleep)``

**Review by tiran — CHANGES_REQUESTED** [review]

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: I've used the same style as the whole unit, do you think I should follow PEP7 in only new function?
I personally prefer that style, but seemed better to follow the code around... the same reasoning I had with the backup.py (but that's completely new...)

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: Ok, will do!

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: Ok.

File: Modules/_sqlite/connection.c
Comment (tiran) [reply Modules/_sqlite/connection.c]: Are you sure? :)
```
+    sqlite3 *bckconn;
+    sqlite3_backup *bckhandle;
```
I guess muscle memory is hard to beat...

Your remark is valid. I don't care that much about PEP 7, as long as you stick to one one style. It's really up to you, either obey PEP 7 for new code or stick to the existing style.

**Comment by lelit** [comment]
@tiran, I've applied your suggestions, except the one about PEP7 style in variable declaration: I personally am -0 on it, because all surrounding code is using a different style.

If you think that's effectively ok I will of course apply that as well. I would even be happy to rectify the whole `connection.c` unit, if that sounds ok.

**Comment by tiran** [comment]
Good work!

Your patch is rather long and touches APIs which I'm not very familiar with. I'll try to get somebody else do a proper review, preferable a Windows dev.

**Comment by lelit** [comment]
Thank you Christian!

W.r.t. PEP7, your example changed my idea and I'm now leaning to use it in the new function: `PyObject* exc, * val, * tb` looks too ugly :-)

File: Modules/_sqlite/connection.c
Comment (tiran) [inline Modules/_sqlite/connection.c]: Sorry, I missed this. We prefer ``PyNone_Check()`` these days.

File: Modules/_sqlite/connection.c
Comment (tiran) [inline Modules/_sqlite/connection.c]: 3.8.7 was released in 2014, https://www.sqlite.org/releaselog/3_8_7.html

Do we need to support older versions? How about requiring a minimum version of 3.8.7 or even newer to compile CPython?

**Review by tiran — CHANGES_REQUESTED** [review]
You have touched the sqlite C code in the last year. Please have a lot, especially at the Windows code.
@berkerpeksag @benjaminp @serhiy-storchaka 

What do you think about requiring a sensible modern version of sqlite to compile the extension module in 3.7?

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [inline Modules/_sqlite/connection.c]: Why not `const char *`?

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [reply Modules/_sqlite/connection.c]: `PyNone_Check()` doesn't exist.

File: Doc/library/sqlite3.rst
Comment (serhiy-storchaka) [inline Doc/library/sqlite3.rst]: I think `*` around `pages` is not needed here.

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: Uhm, really? The only reference I can find is in `Doc/c-api/none.rst` and it says:
```
Note that the :c:type:`PyTypeObject` for ``None`` is not directly exposed in the
Python/C API.  Since ``None`` is a singleton, testing for object identity (using
``==`` in C) is sufficient. There is no :c:func:`PyNone_Check` function for the
same reason.
```
Am I wrong?

File: Doc/library/sqlite3.rst
Comment (serhiy-storchaka) [inline Doc/library/sqlite3.rst]: `':memory:'`

File: Modules/_sqlite/connection.c
Comment (tiran) [reply Modules/_sqlite/connection.c]: Nope, you are right and I got fooled by some legacy code base. I have been banging my head against old C code for the last weeks. The code base has ``#define PyNone_Check(o) ((o) == Py_None)``.

Scratch my comment.

File: Modules/_sqlite/connection.c
Comment (benjaminp) [reply Modules/_sqlite/connection.c]: I'm definitely in favor of not supporting old versions of dependencies if it lets us avoid lots of compatibility code.

Comment (tiran) [reply Modules/_sqlite/connection.c]: According to distrowatch, Debian Jessie has 3.8.7.1 but CentOS 7-1708 only 3.7.17. Guess we need to support < 3.8 for a while.

File: Lib/sqlite3/test/backup.py
Comment (serhiy-storchaka) [inline Lib/sqlite3/test/backup.py]: Does not `unittest.main()` work?

File: Misc/NEWS.d/next/Library/2017-10-05-20-41-48.bpo-27645.1Y_Wag.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2017-10-05-20-41-48.bpo-27645.1Y_Wag.rst]: Period.

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [inline Modules/_sqlite/connection.c]: Is it worth to support path-like objects?

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [inline Modules/_sqlite/connection.c]: `pysqlite_connection_init()` uses a different code for handling errors. Is there an error in `pysqlite_connection_init()` or the code here can be simpler?

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [inline Modules/_sqlite/connection.c]: Leaks the result.

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [inline Modules/_sqlite/connection.c]: Is it worth to make the destination table name configurable?

File: Doc/library/sqlite3.rst
Comment (lelit) [reply Doc/library/sqlite3.rst]: You mean only the first reference to the argument should be emphasized?

File: Lib/sqlite3/test/backup.py
Comment (lelit) [reply Lib/sqlite3/test/backup.py]: Yes, now that the backup suite uses _snake_like_ method names. Done.

File: Misc/NEWS.d/next/Library/2017-10-05-20-41-48.bpo-27645.1Y_Wag.rst
Comment (lelit) [reply Misc/NEWS.d/next/Library/2017-10-05-20-41-48.bpo-27645.1Y_Wag.rst]: Done.

File: Doc/library/sqlite3.rst
Comment (serhiy-storchaka) [reply Doc/library/sqlite3.rst]: I think this isn't a reference to the argument, but a unit. If you want to use a reference to the argument you should write
```
copying up to *pages* pages at a time
```

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: Right, fixed.

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: Why not indeed... but at that point, should I also change the other functions accepting a path name?

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: From my understanding of the [first paragraph](http://www.sqlite.org/c3ref/open.html), I would say yes: `sqlite3_open()` always return a new database connection, with only one exception, OOM.
Do you agree?

If yes, what's the best approach:
a. fix `pysqlite_connection_init()` now
b. defer the fix to a different PR
?

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [reply Modules/_sqlite/connection.c]: On other hand, it may be better to remove support of a file name. The code will be simpler and more reliable if accept only a Connection instance.

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: The second parameter, `"main"`, is the destination _database_ name (see [here](http://www.sqlite.org/c3ref/backup_finish.html#sqlite3backupinit)): in principle yes, it could be exposed as a parameter, but while the source name (ie the fourth parameter) is useful because you may desire to take a snapshot of an _attached_ database, using a target different from `"main"` seems seldom useful...

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: Good catch, fixed!

File: Doc/library/sqlite3.rst
Comment (lelit) [reply Doc/library/sqlite3.rst]: Ok.

File: Doc/library/sqlite3.rst
Comment (lelit) [reply Doc/library/sqlite3.rst]: Fixed.

File: Modules/_sqlite/connection.c
Comment (lelit) [reply Modules/_sqlite/connection.c]: Uhm, yes, _slightly_ simpler indeed, but that would shift the complication to the caller site: in all my use cases I'd use the file name form, so if that gets dropped I'd need to set up another connection instead...
I added the connection form to deal with Cédric's _opposite_ use case, to _restore_ the content of a test database.
I'm -0 on removing such support, but leave the final decision to you.

**Comment by lelit** [comment]
Thank you Serhiy for the review!

Two points remain:

1. should the support for _filename_ be dropped, or should it handle also `pathlib` instances (see [here](https://github.com/python/cpython/pull/4238#discussion_r155725986))?
2. the possible glitch in `sqlite_connection_init()` (see [here](https://github.com/python/cpython/pull/4238#discussion_r155727128))

I will happily follow your directions/decisions.

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [reply Modules/_sqlite/connection.c]: I think a separate issue should be open for fixing `pysqlite_connection_init()`. Since this is a bug fix, it needs to be backported to other branches.

Comment (lelit) [reply Modules/_sqlite/connection.c]: See [#32274](https://bugs.python.org/issue32274): I was wrong about that, the database connection will be released by the `pysqlite_connection_dealloc()` function in any case.

**Comment by lelit** [comment]
@serhiy-storchaka: I have made the requested changes; please review again

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: Do we still need this after a2f15bc?

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: Unfortunately, `sqlite3_errstr` was added in 3.7.15. So we need either add fallback or require SQLite 3.7.15 instead of 3.6.11.

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: It looks like all these header additions can be removed after a2f15bc.

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

File: Modules/_sqlite/connection.c
Comment (serhiy-storchaka) [inline Modules/_sqlite/connection.c]: You can use the `O!` format unit in `PyArg_ParseTupleAndKeywords()`.

**Comment by lelit** [comment]
Thank you, I have made the requested changes; please review again.

**Comment by lelit** [comment]
Is there anything left to do on my part?

**Comment by lelit** [comment]
Will this be merged before 3.7b1?

**Comment by ned-deily** [comment]
@tiran and @berkerpeksag, looks this PR is waiting on your reviews of changes.  Is there anything else holding this up?  It seems like a lot of work went into this PR over a long period and it is quite self-contained.  If everyone agrees it is ready to be merged into master in the next few days and if the buildbots are happy with it, I *might* be persuaded to allow it into 3.7.0b3 as well.

**Comment by berkerpeksag** [comment]
Sorry, I've been busy with family stuff and I will take a look at this PR today or tomorrow.

**Review by berkerpeksag — APPROVED** [review]

**Comment by berkerpeksag** [comment]
This looks pretty good to me. I've rebased @lelit's branch, made some cosmetic changes, and fixed a segfault (https://github.com/python/cpython/pull/4238/commits/824d5f8ded775fc687461414f223c0287eed5bb4)

About supporting a more recent SQLite version as minimum: Last time I've used a newer API (I documented the minimum SQLite version) we got several reports saying "we can't compile Python with SQLite X.Y.Z anymore." Perhaps we can use the death of Python 2.7 as an opportunity and get rid of most of the compatibility code in Python 3.8 or Python 3.9?

**Comment by berkerpeksag** [comment]
@ned-deily

> If everyone agrees it is ready to be merged into master in the next few days and if the buildbots are happy with it, I might be persuaded to allow it into 3.7.0b3 as well.

I'm planning to merge this in an hour or two. I've changed the versionadded marker to 3.8 but should I revert it back to 3.7?

**Comment by ned-deily** [comment]
@berkerpeksag, Thanks for taking this on!  Yeah, let's go crazy and change it to 3.7.  We can always change it later if, for some reason, it doesn't make 3.7.

**Comment by berkerpeksag** [comment]
@ned-deily changed versionadded to 3.7 and added an entry to `Doc/whatsnew/3.7.rst`. What's the current process for getting something in after b1? Do I need to open a PR against the 3.7 branch or some special branch?

**Comment by ned-deily** [comment]
@berkerpeksag, you can just the "need backport to 3.7" label to start an auto backport to 3.7

**Comment by berkerpeksag** [comment]
@lelit thank you for your patience! I know that this was a frustrating experience for you, but I hope you are going to continue contributing to Python :)

**Comment by lelit** [comment]
@berkerpeksag thank you and don't worry, I've been around for enough time to know how it goes and I will surely keep that for the remaining half of my life :smiley:

**Pull request merged**
